### ISSUE TITLE
Transformer实验 | J球星的博客

### ISSUE THREAD

**Issue by weiren1998** [weiren1998/weiren1998.github.io#13]
https://weiren1998.github.io/archives/697052e2.html#more 

 在探索网络架构的过程中，需要做很多尝试和思考，同时也需要把实验数据和对于结果的思考等记录下来，从而一点点积累感觉
 
 
 最近在做本科的毕业设计，题目是足球视频中的行为关键帧检测算法设计。在实验过程中，发现有很多细小的想法，但有时一晃而过，可能是一些小的尝试，但很少会做对比试验，因此将这些想法记录下来，方便之后再做改进。

**Comment by weiren1998** [comment]
#### 测试极大值对于bn和ln的影响

- 当所有patch的同一个channel数值很大时，过BN后同一个patch的不同channel之间会比较接近，均接近0；但过LN后同一个patch的那一个channel的值依旧很大，且所有patch的那个channel数值都很大
- 当所有batch中，一个patch维度的所有channel值比其他所有patch的cahnnel值都大很多时，过BN后，该patch的所有channel还是会很大；LN会让他们变得很小

|     Model     | ImNet top-1 (%) |  Para (M)  | Throughput  (img/s) |  Embed dim  |  #heads   | #depth |
| :-----------: | :-------------: | :--------: | :-----------------: | :---------: | :-------: | :----: |
|    Deit-Ti    |      72.2       |     5      |       2536.5        |     192     |     3     |   12   |
|    Deit-S     |      79.8       |     22     |        940.4        |     384     |     6     |   12   |
|    Deit-B     |      81.8       |     86     |        292.3        |     768     |    12     |   12   |
| <u>Deit-L</u> |                 | <u>153</u> |                     | <u>1024</u> | <u>16</u> |   12   |

**Comment by weiren1998** [comment]
#### 细品BN的中间输出值，将其可视化
~~~python
import cv2
import numpy as np
img = x.cpu().numpy() # 将torch变为numpy
img = (img - np.min(img))/(np.max(img) - np.min(img)) *255.0 # 归一化*255
img = img.astype(int) # 转为int
cv2.imwrite('./100_msa.jpg', img[100]) # 将一个feature输出
~~~

**Comment by weiren1998** [comment]
### 3. Deit-s 换成bn后为什么会崩？

猜想：

1. **channel变大后，channel一个维度很大情况下，bn不能把他消下去，而ln可以。**
   1. 验证：在tiny上去改为bn试试，然后尝试在tiny和B之间的channel dim=288=(192+384)/2
   2. 验证：在各block之间加hook存一下channel维度的值，看其大小
2. 为什么要换成bn：bn在下游检测任务中有提升







- bn模型layer1的输入：

  ![gelu](C:\Users\r50021608\AppData\Roaming\Typora\typora-user-images\image-20220505174533602.png)

  ![relu6](C:\Users\r50021608\AppData\Roaming\Typora\typora-user-images\image-20220505174912093.png)

- ln模型layer1的输入：

  ![gelu](C:\Users\r50021608\AppData\Roaming\Typora\typora-user-images\image-20220505174748965.png)

  ![relu6](C:\Users\r50021608\AppData\Roaming\Typora\typora-user-images\image-20220505175038382.png)

- bn模型layer11的输出：

  ![gelu](C:\Users\r50021608\AppData\Roaming\Typora\typora-user-images\image-20220505174616028.png)

  ![relu6](C:\Users\r50021608\AppData\Roaming\Typora\typora-user-images\image-20220505174959407.png)

- ln模型layer11的输出：

  ![gelu](C:\Users\r50021608\AppData\Roaming\Typora\typora-user-images\image-20220505174816488.png)

  ![relu6](C:\Users\r50021608\AppData\Roaming\Typora\typora-user-images\image-20220505175210690.png)

#### 目的是提出一个新的归一化方法，结合BN和LN的优点





- 实验：
  - load中间模型，可视化中间层结果

    - 先看一下可以完成训练的ln模型的feature map：ln的feature map中C的某个维度有极大值和极小值，原因是ln中有weight过大和过小

      ~~~python
      # base ln epoch 50 layer1
      (Pdb) self.norm1.weight.min()
      tensor(-0.0013, device='cuda:0')
      (Pdb) self.norm1.weight.max()
      tensor(0.3647, device='cuda:0')
      (Pdb) self.norm2.weight.max()
      tensor(1.5141, device='cuda:0')
      (Pdb) self.norm2.weight.min()
      tensor(0.0197, device='cuda:0')
      ~~~

      

  - 

|   Model    |  BN  |    ImNet top-1 (%)     | Para (M) |      Throughput  (img/s)       | Embed dim | #heads | #depth |
| :--------: | :--: | :--------------------: | :------: | :----------------------------: | :-------: | :----: | :----: |
|  Deit-Ti   |  ×   |      72.56 (72.2)      |    5     |             2536.5             |    192    |   3    |   12   |
|   Deit-S   |  ×   |      80.47 (79.8)      |    22    |             940.4              |    384    |   6    |   12   |
|   Deit-B   |  ×   |         (81.8)         |    86    |             292.3              |    768    |   12   |   12   |
|   Deit-L   |  ×   |      <u>77.83</u>      |  153.2   |                                |   1024    |   16   |   12   |
| Deit-Ti-bn |  √   |         70.26          |   5.7    |                                |    192    |        |        |
| Deit-S-bn  |  √   | 77.14 (<u>236 epo</u>) |   22.1   | 237epoch acc掉的厉害，看看原因 |    384    |        |        |
| Deit-B-bn  |  √   | 48.14 (<u>13 epo</u>)  |   86.6   |                                |    768    |        |        |
| Deit-L-bn  |  √   |  41.88 (<u>9 epo</u>)  |  153.2   |                                |   1024    |        |        |


#### 测试极大值对于bn和ln的影响

- 当所有patch的同一个channel数值很大时，过BN后同一个patch的不同channel之间会比较接近，均接近0；但过LN后同一个patch的那一个channel的值依旧很大，且所有patch的那个channel数值都很大
- 当所有batch中，一个patch维度的所有channel值比其他所有patch的cahnnel值都大很多时，过BN后，该patch的所有channel还是会很大；LN会让他们变得很小

**Comment by weiren1998** [comment]
#### 3. 尝试训练Deit_bn模型，使得训练稳定

##### 3.1 将activation function换为relu6

1. 动机：正常训练时梯度会爆炸，因此将FFN中的激活层变为relu6，减小激活层的输出值，防止正向传播时特征爆炸，同时可以把较大的x对应的梯度降为0

##### 3.2 加入Weight Norm机制

1. 动机：weight norm貌似可以在CNN中稳定训练（具体可以看看论文）

##### 3.3 batchsize的大小与BN的稳定性好像也有关

1. Deit-L_bn_wn模型，当batchsize为128时，模型在第12个epoch就NaN了，而当把batchsize调整为32后，模型就可训练了
2. **减小batchsize好像可以使BN训练更稳定，有必要在这个方向继续挖么？** 

| No.  |         Model         |  BN  | ImNet top-1 (%) | Para (M) |      Throughput  (img/s)      | Embed dim | #heads | #depth |
| :--: | :-------------------: | :--: | :-------------: | :------: | :---------------------------: | :-------: | :----: | :----: |
|      |        Deit-Ti        |  ×   |  72.56 (72.2)   |    5     |            2536.5             |    192    |   3    |   12   |
|      |        Deit-S         |  ×   |  80.47 (79.8)   |    22    |             940.4             |    384    |   6    |   12   |
|      |        Deit-B         |  ×   |  78.74 (81.8)   |   86.6   |             292.3             |    768    |   12   |   12   |
|      |        Deit-L         |  ×   |      77.83      |  153.2   |                               |   1024    |   16   |   12   |
|  1   |    Deit-B-bn_relu6    |  √   |      79.26      |   86.6   |                               |    768    |   12   |   12   |
|  2   |    Deit-L-bn_relu6    |  √   |                 |  153.2   |          看会不会崩           |           |        |        |
|  3   |     Deit-B-bn_wn      |  √   |    priv ing     |   86.7   |       看wn模块是否有用        |    768    |   12   |   12   |
|  4   |     Deit-L-bn_wn      |  √   |     BJ ing      |  153.3   |      看看wn会不会崩的快       |           |        |        |
|  5   |  Deit-B-bn_wn_relu6   |  √   |    SH BJ ing    |   86.7   | 看relu6和wn一起是否有好的效果 |           |        |        |
|  6   |    Deit-B_ln_bn_wn    |  √   |   ipt BJ ing    |   86.7   |     看LN+BN+WN能不能训练      |           |        |        |
|  7   |  Deit-B_ln_bn_relu6   |  √   |   NaN (8 epo)   |   86.7   |          查找NaN原因          |           |        |        |
|  8   | Deit-B_ln_bn_wn_relu6 |  √   |   ipt BJ ing    |   86.7   |  看LN+BN+WN+relu6能不能训练   |           |        |        |
|      |    Swin-S_bn_relu6    |  √   |     ipt ing     |   49.7   |       看是否适用于swin        | batch 256 |        |        |
|      |    Swin-T_bn_relu6    |  √   |     ipt ing     |   28.3   |    准备在检测任务上看效果     | batch 256 |        |        |



### 4. TODO

1. 将LN-BN模型加入WN (-+relu6) 来实验 √
2. Deit-B的结果比论文低很多，用Deit原始的轮子训练看所有模型的结果 
3. 测试推理速度，测试推理时的throughput，比较Deit-B和Deit-B_bn
4. 测试Swin-T_bn_relu6和Swin-T_bn_wn在**下游任务**（检测&分割）上的性能
5. 探究ln-bn-relu6模型为何会崩掉，relu6对于稳定训练的帮助到底是什么

**Comment by weiren1998** [comment]
Transformer-based architectures have achieved great success in Natural Language Processing (NLP) and Computer vision (CV). It was first proposed in the NLP tasks. In previous Convolutional Neural Networks (CNNs), the Batch Normalization (BN) is a widely-adopted method to reduce internal covariate shift and improve the generalization of the neural networks. While the Layer Normalization (LN) is commonly used in Recurrent Neural Networks (RNNs) and Transformers. When the Transformer was applied to CV tasks, it simply follow the original setting in NLP tasks by using LN because of the unstable training with BN. However, BN may be more proper to handle the image data and it could accelerate the inference process by using the moving average and variance. In this paper, we figure out that BN will make the variance between channels become very large, so we try to replace the original GeLU with a truncated ReLU6, and we also use the Weight Normalization (WN) to make the training more stable. Our experiments show that the BN-based Transformer model could have a comparable result with the LN-based model, and the BN-based model could speed up inference time by 12%.
